src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -41,7 +47,9 @@ def get_preference(
 
     def find_matches(self, requirement):
         # type: (Requirement) -> Sequence[Candidate]
-        return requirement.find_matches()
+        constraints = self._constraints.get(requirement.name, [])
+        is_root = (requirement.name in self._roots)
+        return requirement.find_matches(is_root, constraints)
```

Comment:
(This is rambling, I don’t really mind that much) I don’t name variables this way if I can help it because `is_blah` reads like a function.

Reply:
I wanted to avoid `root` because I'm massively over-using that name. But I don't mind changing it. I don't have the energy to come up with something better, though...

Reply:
`i_am_root` :)

![groot](https://user-images.githubusercontent.com/3275593/81132271-71b23b00-8f6b-11ea-9727-9e546284377d.gif)

Reply:
Here's another color for the bikeshed:

```suggestion
        return requirement.find_matches(
            user_specified=(requirement.name in self._roots),
            constraints=self._constraints.get(requirement.name, []),
        )
```

Reply:
Counting this as resolved for now. We can revisit this if needed in a follow up PR.

Reply:
Yeah, this will have massive conflicts. I might resurrect it if I can merge, or just start a new PR. It's purely about moving `root_reqs` off the factory now.